### PR TITLE
Increase Test Coverage, Updated Doc, Feature Flags and Reduce Logging

### PR DESCRIPTION
**Pull request by grabnerandi**
This PR includes the following changes
1. Increased Test coverage as raised in #78 
2. Implemented new dynatrace.conf.yaml option to control dashboard query feature as discussed in #82 
3. Updated readme.md with more details on different use cases and explained dynatrace.conf.yaml
4. Reduced some unnecessary logging. Here though more log cleanup would be possible but this should be implemented using proper log level handling as described here - https://github.com/keptn/keptn/issues/2413
5. Only parsing dashboard if dashboards have changed from prevuios runs - see #92 
6. Updated default set of SLIs to use Metric V2 API - see #91 
7. Allow placeholders in dtCreds - see #90

### CODE REVIEW

File: pkg/lib/dynatrace/dynatrace.go [hunk 1]
```diff
@@ -1103,55 +1133,125 @@ func (ph *Handler) QueryDynatraceDashboardForSLIs(project string, stage string,
...
-
-	// now we are enriching it with all the additonal parameters, e.g: time, filters ...
-	metricsQuery, metricID := ph.BuildDynatraceMetricsQuery(metricsQuery, startUnix, endUnix, customFilters)
-
-	ph.Logger.Debug(fmt.Sprintf("trying to fetch metric %s", metricID))
-	result, err := ph.ExecuteMetricsAPIQuery(metricsQuery)
-
-	if err != nil {
-		return 0, fmt.Errorf("error from Execute Metrics API Query: %s\n", err.Error())
+		return 0, fmt.Errorf("Error when fetching SLI config for %s %s\n", metric, err.Error())
```
[the remark this answers is not in the document]
Comment (christian-kreuzberger-dtx) [reply pkg/lib/dynatrace/dynatrace.go:1163]: ignoring this golint warning

File: pkg/lib/dynatrace/dynatrace.go [hunk 2]
```diff
@@ -1103,55 +1128,125 @@ func (ph *Handler) QueryDynatraceDashboardForSLIs(project string, stage string,
...
+		//
+		// In this case we are querying regular MEtrics
+		// now we are enriching it with all the additonal parameters, e.g: time, filters ...
+		metricsQuery, metricID := ph.BuildDynatraceMetricsQuery(metricsQuery, startUnix, endUnix)
+		result, err := ph.ExecuteMetricsAPIQuery(metricsQuery)
 
-				actualMetricValue = i.Data[0].Values[0]
-				break
+		if err != nil {
+			return 0, fmt.Errorf("error from Execute Metrics API Query: %s\n", err.Error())
```
[the remark this answers is not in the document]
Comment (christian-kreuzberger-dtx) [reply pkg/lib/dynatrace/dynatrace.go:1238]: ignoring this golint warning

File: pkg/lib/dynatrace/dynatrace.go [hunk 3]
```diff
@@ -1103,55 +1128,125 @@ func (ph *Handler) QueryDynatraceDashboardForSLIs(project string, stage string,
...
+		tileName := querySplits[1]
+		requestedDimensionName := querySplits[2]
+		usqlRawQuery := querySplits[3]
+
+		usql := ph.BuildDynatraceUSQLQuery(usqlRawQuery, startUnix, endUnix)
+		usqlResult, err := ph.ExecuteUSQLQuery(usql)
+
+		if err != nil {
+			return 0, fmt.Errorf("Error executing USQL Query %v", err)
+		} else {
```
[the remark this answers is not in the document]
Comment (christian-kreuzberger-dtx) [reply pkg/lib/dynatrace/dynatrace.go]: @grabnerandi Is this correct? Please take a look and drop the else block, if it's correct.

File: pkg/lib/dynatrace/dynatrace.go [hunk 4]
```diff
@@ -292,75 +324,75 @@ func (ph *Handler) findDynatraceDashboard(project string, stage string, service
...
 	}
-	defer resp.Body.Close()
-
-	ph.Logger.Debug("Request finished, parsing body...")
-
-	body, _ := ioutil.ReadAll(resp.Body)
-
-	ph.Logger.Debug(string(body) + "\n")
+	if resp == nil || resp.StatusCode != 200 {
+		return nil, fmt.Errorf("No valid response from metrics api!")
```
[the remark this answers is not in the document]
Comment (christian-kreuzberger-dtx) [reply pkg/lib/dynatrace/dynatrace.go:394]: Ignore this

**Review by bacherfl — APPROVED** [review]
lgtm

**Comment by agrimmer** [comment]
Does this PR implement https://github.com/keptn-contrib/dynatrace-sli-service/issues/90?
If so, with this feature a Keptn project-stage-service combination would not be mapped to one DT tenant anymore. Right?
Can you please elaborate on the use case?

**Comment by grabnerandi** [comment]
Hi @agrimmer - yes - this also implements #90 but I dont think it changes anything from before. Before you could define the secret name the dynatrace-sli-service should be using in dtCreds as a single string. Now we are simply allowing you to also use the the same placeholders we allow you to use in SLI definitions. It would allow you to map to different tenants - but - the same is true if you would simply change the values in the dynatrace-secret between quality gate runs.

**Review by agrimmer — CHANGES_REQUESTED** [review]
Clarify https://github.com/keptn-contrib/dynatrace-sli-service/issues/90

**Comment by agrimmer** [comment]
Am I right that now labels would allow to dynamically select different DT-tenants for one Keptn service?
If so, we have no fixed mapping anymore between a Keptn service in a stage and project to a DT-tenant.
We at least have to think about the consequences first.

**Comment by grabnerandi** [comment]
HI @agrimmer - you are correct. Technically it is possible to switch between different tenants on the same project/stage/service combination if you are using e.g: $LABELs. There are use cases where it makes sense to use $LABEL.xxx as placeholder but as a general best practice I suggest to use e.g: $STAGE as placeholder. Here is an example
```
dtCrends: dynatrace-$STAGE
```

This dynatrace.conf.yaml will - depending on the stage - pick the correct dynatrace credentials, e.g: dynatrace-dev, dynatrace-staging, dynatrace-prod ... -> this is a very common use case and can now be easily implemented with these placeholders and eliminates the need to upload different dynatracce.conf.yaml's for each stage

**Review by agrimmer — APPROVED** [review]
@grabnerandi thanks for the explanations!

**Comment by agrimmer** [comment]
@grabnerandi Can you please close all issues which have been resolved with this PR?

**Pull request merged**